### PR TITLE
Modify how quotations handle whole sentences.

### LINKED ISSUE #13091: Sentence splitting severely broken in Coqide 8.12.0
I have encountered several cases where Coqide is completely unusable in 8.12.0. In the example below, Coqide considers that the input file stops at `Proof`. It completely ignores what follows. In some other cases, it just ignores some sentences (which in turn can cause the wrong tactics to be applied to the wrong goals, hence infinite loops, @MSoegtropIMC).

```coq
Goal True.
Proof.
let vars := constr:((0)%nat) in
exact I.
Qed.
```

This is a severe regression with respect to 8.11.

### PR DESCRIPTION
**Pull request by silene**
Instead of reverting #10394, this pull request modifies Coq's lexer so that `:{{` is always a sentence-gobbling quotation, whether registered or not, and so that `:(` and `:[` never gobble sentences, even if they are registered as quotations.

This should make Elpi usable in Coqide (though I did not test).

Fixes / closes #13091

### CODE REVIEW

**Comment by silene** [comment]
The failures in the testsuite looks like that:
```diff
@@ -1,5 +1,5 @@
 
-Coq < 
+Coq < Coq < 
 Coq < Coq < 1 subgoal
   
   ============================
@@ -88,4 +88,4 @@
 
 p14 < 
 Coq < 
-Coq < 
+
```

In other words, some prompts are displayed earlier. Presumably, this is due to `Stream.npeek 3 s`. Indeed, since Coq is now looking a bit more ahead of the character stream, newlines are encountered earlier, and thus  prompt are displayed earlier. Should we care?

**Comment by gares** [comment]
This approach is fine as well.
Thanks for looking into this

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
I can test this today

**Comment by gares** [comment]
It works great!

**Review by gares — APPROVED** [review]

**Comment by ejgallego** [comment]
That feels like quite a change for a point release (cc: @Zimmi48 ) 

@herbelin @ppedrot and miscellaneous @coq/parsing-maintainers , what do you think?

**Comment by Zimmi48** [comment]
Agreed, I would merge the other PR (the revert) for v8.12 if this is the solution chosen for master.

**Comment by gares** [comment]
OK, I changed the milestone then. The advantage of this PR w.r.t. the revert is that elpi works fine, so I'd like to merge this in master. Waiting for a review by @coq/coqide-maintainers

**Comment by herbelin** [comment]
Both solutions are ok to me for 8.12.1 (only revert or also `:{{` - though, to be honest, I did not study the Elpi need). A fortiori, `:{{` in master is ok (with the same remark that I did not study the exact Elpi requirements).

**Comment by gares** [comment]
elpi needs a way to delimit arbitrary text. I usually use `lp:{{...}}` but when I wrote the patch I also added support for `xxx:((...))` and `xxx:[[...]]` for `xxx` a registered quotation-keyword, trying to provide a service to other extension/embedded languages. My silly mistake I did not foresee that `xxx:((` could conflict with common expressions and that CoqIDE cannot easily be informed about the `xxx` which are registered. I should have probably added `lp:{{` only.

This PR makes CoqIDE recognize any `xxx:{{`, which is more than enough for elpi. The revert, well, it is also reverts support for elpi. So I'm happy with this PR in master. For the point release, I'm OK with any choice.

**Comment by Zimmi48** [comment]
OK, after reading again the explanations in this PR and thinking a bit, it looks like the following situation:
- this is the first time that PR #10394 has made it to a final release (it had always been reverted in some beta release before);
- therefore, there is new support for Elpi quotations in 8.12;
- therefore, reverting it would actually introduce a feature regression in 8.12.1;
- this PR instead limits the feature to the only case that is known to be used (so it's a more reasonable feature regression) while still fixing the CoqIDE bug.

Therefore, I've changed my mind and I am in favor of accepting this PR for 8.12.1 and closing the revert PR if @ejgallego agrees.

PS: #13094 reverting #10394, easy to get confused with just two digits swapped!

**Comment by gares** [comment]
Since Hugo is ok with this pr, I ask @coqbot merge now

**Comment by herbelin** [comment]
> elpi needs a way to delimit arbitrary text.

To add a comment here. There is	similar need with custom notations where users would like that what is in the scope of a custom grammar does not interfere (especially tokens) with the constr side.

This is difficult to do with the current parsing model and I'm thinking to various solutions such as identifying a class of pairs of delimitors (such as @gares' one) which the constr parser would not try to interpret, nor even to lex. Square brackets and curly braces seem good candidates (but I start to worry with people working with asymmetric parenthesizing such as `|01>` for qubits...).

**Pull request merged**
